src/main/kotlin/org/pkl/lsp/VirtualFile.kt

```diff
@@ -75,19 +79,58 @@ interface VirtualFile : ModificationTracker {
   /** Tells if this file is within a package. */
   val `package`: PackageDependency?
 
+  var contents: String
+
   fun parent(): VirtualFile?
 
   fun resolve(path: String): VirtualFile?
 
-  fun toModule(): PklModule?
-
-  fun contents(): String
+  fun getModule(): CompletableFuture<PklModule?>
 }
 
 sealed class BaseFile : VirtualFile {
-  internal val modificationCount: AtomicLong = AtomicLong(0)
+  override fun getModificationCount(): Long = version ?: -1L
+
+  override var version: Long? = null
+
+  abstract fun doReadContents(): String
 
-  override fun getModificationCount(): Long = modificationCount.get()
+  // Use if contents have not yet been set, read contents from external source, possibly causing
+  // I/O.
+  override var contents: String
+    get() {
+      return myContents ?: doReadContents()
+    }
+    set(text) {
+      myContents = text
+    }
+
+  final override fun getModule(): CompletableFuture<PklModule?> {
+    return project.cachedValuesManager.getCachedValue("VirtualFile($uri)") {
+      val future = CompletableFuture.supplyAsync { doBuildModule() }
+      CachedValue(future, this)
+    }!!
+  }
```

Comment:
Using the cached values API with futures means that if the file is still valid (i.e. its contents have not changed), it will return the same future, which means we only build the module once.

The virtual file itself is the cache dependency; updating the virtual file will invalidate the cache, and a new future is computed.